### PR TITLE
Fix for e2e tests

### PR DESCRIPTION
**Pull request by horheynm**
SUMMARY:
Issue was caused by trying to save the recipe on modify_save_pretrained. This runs after the session is reset, so no recipe to save. 

With SparseAutoModelForCausalLM before the commit, in the same test, recipe is not saved. It is saved in the save_model_and_recipe, which runs after oneshot. We dont do it here anymore because 1.  if output path is not provided it doesnt save; 2. if we add again, there will be double save

[Notes]
Another way to fix this is to not to reset the session. This allows us to load and save the recipe. pre-commited SparseAutoModelForCausalLM state does not save the recipe.

TEST PLAN:
Checked that the output from oneshot is the same and pass the e2e tests.

### CODE REVIEW

**Review by dsikka — COMMENTED** [review]
Why is the session being reset in these cases and not our example cases? I'm just surprised we didn't see it then

File: src/llmcompressor/core/session.py [hunk 1]
```diff
@@ -260,12 +262,16 @@ def reset_stage(self):
...
         """
         :return: serialized string of the current compiled recipe
         """
         recipe = self.lifecycle.recipe_container.compiled_recipe
-        return recipe.yaml()
+
+        if recipe is not None and hasattr(recipe, "yaml"):
+            return recipe.yaml()
+
+        logger.warning("Recipe not found in session - may been reset")
```
Comment (dsikka) [inline src/llmcompressor/core/session.py:274]: grammar

Comment (dsikka) [reply src/llmcompressor/core/session.py:274]: I also think we should be saving the recipe for this case.

Comment (horheynm) [reply src/llmcompressor/core/session.py:274]: It doesnt save there bc there is not recipe bc session was reset

Comment (horheynm) [reply src/llmcompressor/core/session.py:274]: So the recipe is not saved using SparseAutoModelForCausalLM anyways, Yes i do think we should be saving the recipes but saving/ not saving do not affect the outcome of this test

Comment (dsikka) [reply src/llmcompressor/core/session.py:274]: Do you mind updating it in this PR to save it then?

Comment (horheynm) [reply src/llmcompressor/core/session.py:274]: Good question, 
the save happens after those lines at the context manager. 
```python3
recipe_path = os.path.join(save_directory, "recipe.yaml")
            session = active_session()
            recipe_yaml_str = session.get_serialized_recipe()
            with open(recipe_path, "w") as fp:
                fp.write(recipe_yaml_str)
```
The problem we have now is that we get an error at `recipe_yaml_str = session.get_serialized_recipe()`. 
Reason is because it assumes that there is an object, which can be serialized into yaml. The object is None in this case because the session is reset. There is nothing there -> None.

Comment (dsikka) [reply src/llmcompressor/core/session.py]: my question is why it's being reset now and not before to trigger this error

Comment (horheynm) [reply src/llmcompressor/core/session.py]: Its on the description

```
With SparseAutoModelForCausalLM before the commit, in the same test, recipe is not saved. It is saved in the save_model_and_recipe, which runs after oneshot. We dont do it here anymore because 1. if output path is not provided it doesnt save; 2. if we add again, there will be double save

```

tldr: it follows a differnt save pathway

Comment (horheynm) [reply src/llmcompressor/core/session.py]: The session has always been reset, it just doesnt save the recipe 
https://github.com/vllm-project/llm-compressor/pull/927/files#diff-a863d6754eda15d0560a67cbe262489d9784b7c0ebf8b1f43cf6d17389b85fc9R135

**Review by dsikka — CHANGES_REQUESTED** [review]
Could we update such that we're saving the recipe now please?
It's informative to have that for the e2e cases.

**Comment by horheynm** [comment]
@dsikka 
The recipe now should be saved.
FYI for this e2e tests, saving recipes do not change the outcome so it wont be informative.

**Comment by dsikka** [comment]
> @dsikka The recipe now should be saved. FYI for this e2e tests, saving recipes do not change the outcome so it wont be informative.

We push the outputs and refer to the frequently. Please look at the complete test

**Comment by horheynm** [comment]
> > @dsikka The recipe now should be saved. FYI for this e2e tests, saving recipes do not change the outcome so it wont be informative.
> 
> We push the outputs and refer to them frequently. Please look at the complete test

If we were referring to it frequently, we should have known recipe wasnt being pushed previously. 
I can add recipe creation check on the test in the follow up pr.

**Review by dsikka — APPROVED** [review]

**Review by rahul-tuli — APPROVED** [review]

**Comment by horheynm** [comment]
dont merge yet please, not resetting session, so lifecycle is finializing more than once, which raises error.

**Review by kylesayrs — APPROVED** [review]

**Comment by horheynm** [comment]
Follow up pr to check the contexts of the files, and to make sure recipes exist in the session is here 

https://github.com/vllm-project/llm-compressor/pull/929

**Review by robertgshaw2-neuralmagic — APPROVED** [review]

**Review by mgoin — APPROVED** [review]
